File: 1/AI1110/Random-Numbers/codes/7.1.c
// Writes to a file the probabilities P_e for different values of gamma

// Name: Ankit Saha
// Roll number: AI21BTECH11004

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "header.h"

#define NUM 1000000

int main() {
	FILE *fp = fopen("7.1.dat", "w");
	for (double gamma = 0; gamma <= 10; gamma += 0.25) {
		int e = 0;
		int total = 0;
		double sigma = sqrt(gamma/2);
		generateY_Rayleigh("7.1.y.dat", NUM, sigma);
		double *y_data = readData("7.1.y.dat", NUM);
		double *x_data = readData("7.1.x.dat", NUM);
		for (int i = 0; i < NUM; i++) {
			if (y_data[i] < 0 && x_data[i] > 0) {
				e++;
			}
			if (x_data[i] > 0) {
				total++;
			}
		}	
		double p_e = (double)e / total;
		fprintf(fp, "%lf\n", p_e);
	}
	fclose(fp);
	return 0;	
}

File: 1/AI1110/Random-Numbers/codes/7.1.py
# Plots theoretical and empirical P_e with respect to gamma from 0 to 10 dB

# Name: Ankit Saha
# Roll number: AI21BTECH11004

import numpy as np
import matplotlib.pyplot as plt
import mpmath as mp
import scipy

p_e = np.loadtxt('7.1.dat', dtype='double')
NUM = len(p_e)
gamma = np.linspace(0, 10, 41)

def P_e_Rayleigh(gamma):
    tmp = gamma/(gamma + 2)
    return 0.5 * (1 - tmp**0.5)

vec_p_e = scipy.vectorize(P_e_Rayleigh, otypes=[float])

plt.plot(gamma, vec_p_e(gamma), label='Theoretical')
plt.plot(gamma, p_e, 'o', label='Empirical')
plt.grid()
plt.legend()
plt.xlabel('$\gamma$')
plt.ylabel('$P_e$')
plt.savefig('../figs/7.1.png')
plt.show()
